libtenzir/include/tenzir/split_at_regex.hpp

```diff
@@ -0,0 +1,75 @@
+//    _   _____   __________
+//   | | / / _ | / __/_  __/     Visibility
+//   | |/ / __ |_\ \  / /          Across
+//   |___/_/ |_/___/ /_/       Space and Time
+//
+// SPDX-FileCopyrightText: (c) 2025 The Tenzir Contributors
+// SPDX-License-Identifier: BSD-3-Clause
+
+#pragma once
+
+#include "tenzir/chunk.hpp"
+#include "tenzir/generator.hpp"
+
+#include <boost/regex.hpp>
+
+#include <optional>
+#include <string_view>
+
+namespace tenzir {
+
+inline auto split_at_regex(std::string_view separator) {
+  auto expr = boost::regex{separator.data(), separator.size(),
+                           boost::regex_constants::no_except
+                             | boost::regex_constants::optimize};
+  return [exp = std::move(expr)](generator<chunk_ptr> input) mutable
+           -> generator<std::optional<std::string_view>> {
+    auto expr = std::move(exp); // NOLINT
+    auto buffer = std::string{};
+    auto consumed = true;
+    boost::match_results<std::string::const_iterator> what;
+    for (auto&& chunk : input) {
+      if (not chunk or chunk->size() == 0) {
+        co_yield std::nullopt;
+        continue;
+      }
+      buffer.append(reinterpret_cast<const char*>(chunk->data()), chunk->size());
+      const auto first = buffer.cbegin();
+      auto current = first;
+      auto begin = current + (consumed ? 0 : 1);
+      while (begin <= buffer.end()
+             and boost::regex_search(begin, buffer.cend(), what, expr)) {
+        // Don't yield when we reached the end as a longer match could be found
+        // with the subsequent characters at the beginning of the next chunk.
```

Comment:
This has one additional problem that probably warrants a timeout: In a never-ending stream, say we're receiving bytes over TCP, you may run into a situation where the last event is withheld if there's a valid separator at the end.

Reply:
> you may run into a situation where the last event is withheld if there's a valid separator at the end.

That is only the case when the connection stays open. Changing behavior based on timing is not a good idea and I'm quite opposed to implementing that.

Reply:
Please do create a follow-up issue. We had customer reports before that this behavior was broken for Syslog, and this has very much the same symptoms. Not creating a story and just resolving the thread goes against the spirit of "approved with comments."